contracts/V1/StateChannelDiamondProxy/FraudProofFacet.sol

```diff
@@ -47,448 +43,251 @@ contract FraudProofFacet is StateChannelCommon {
-        // Implementation for handling DisputeNotLatestState proof
-        return address(0); // Replace with actual logic
+        // check signatures
+        address signer = StateChannelUtilLibrary.retriveSignerAddress(
+            thresholdBlockConfirmation.signedBlock.encodedBlock,
+            thresholdBlockConfirmation.signedBlock.signature
+        );
+        address[] memory signers = _collectBlockConfirmationAddresses(
+            thresholdBlockConfirmation.signedBlock.encodedBlock,
+            thresholdBlockConfirmation.signatures
+        );
+
+        if(signer != msg.sender || !StateChannelUtilLibrary.isAddressInArray(signers, msg.sender)) {
+            return address(0);
+        }
+        // If calldata check also fails, return false with the last error message
+        return originalTimedOutDispute.disputer;
     }
 
-    function _handleDisputeInvalid(bytes memory encodedProof, FraudProofVerificationContext memory fraudProofVerificationContext) internal returns (address) {
-        // Implementation for handling DisputeInvalid proof
-        return address(0); // Replace with actual logic
-    }
+    function _handleTimeoutPriorInvalid(bytes memory encodedProof, FraudProofVerificationContext memory fraudProofVerificationContext) view internal returns (address) {
+        TimeoutPriorInvalidProof memory timeoutPriorInvalidProof = abi.decode(encodedProof, (TimeoutPriorInvalidProof));
+        Dispute memory originalDispute = timeoutPriorInvalidProof.originalDispute;
+        Dispute memory recursiveDispute = timeoutPriorInvalidProof.recursiveDispute;
 
-    function _handleDisputeInvalidRecursive(bytes memory encodedProof, FraudProofVerificationContext memory fraudProofVerificationContext) internal returns (address) {
-        // Implementation for handling DisputeInvalidRecursive proof
-        return address(0); // Replace with actual logic
+        if(recursiveDispute.channelId != originalDispute.channelId && recursiveDispute.channelId != fraudProofVerificationContext.channelId) {
+            return address(0);
+        }
+        // check if the recursive dispute is available
+        bytes32 recursiveDisputeCommitment = keccak256(abi.encode(
+            recursiveDispute,
+            timeoutPriorInvalidProof.recursiveDisputeTimestamp
+        ));
+        bytes32 originalDisputeCommitment = keccak256(abi.encode(
+            originalDispute,
+            timeoutPriorInvalidProof.originalDisputeTimestamp
+        ));
+
+        (bool isAvailable, bytes32 commitment) = getDisputeCommitment(fraudProofVerificationContext.channelId, recursiveDispute.disputeIndex);
+
+        if(!isAvailable && commitment != recursiveDisputeCommitment) {
+            return address(0);
+        }
+        if(recursiveDispute.previousRecursiveDisputeIndex == type(uint256).max){
```

Comment:
This is ok - you check if it's not a recursive dispute, but if it is you also need to check that recursiveDispute.previousRecursiveDisputeIndex == originalDispute.disputeIndex